dotnet/src/SemanticKernel.Core/Functions/KernelFunctionFromPrompt.cs

```diff
@@ -228,5 +245,59 @@ private async Task<(ITextGenerationService, string, PromptRenderedEventArgs?)> R
     /// <summary>Create a random, valid function name.</summary>
     private static string RandomFunctionName() => $"func{Guid.NewGuid():N}";
 
+    /// <summary>
+    /// Captures usage details, including token information.
+    /// </summary>
+    private void CaptureUsageDetails(string? modelId, IDictionary<string, object?>? metadata, ILogger logger)
+    {
+        if (string.IsNullOrWhiteSpace(modelId))
+        {
+            logger.LogWarning("No model ID provided to capture usage details.");
+            return;
+        }
+
+        if (metadata is null)
+        {
+            logger.LogWarning("No metadata provided to capture usage details.");
+            return;
+        }
+
+        if (!metadata.TryGetValue("Usage", out object? usageObject) || usageObject is null)
+        {
+            logger.LogWarning("No usage details provided to capture usage details.");
+            return;
+        }
+
+        var promptTokens = 0;
+        var completionTokens = 0;
+        try
+        {
+            var jsonObject = JsonSerializer.Deserialize<JsonElement>(JsonSerializer.Serialize(usageObject));
+            promptTokens = jsonObject.GetProperty("PromptTokens").GetInt32();
+            completionTokens = jsonObject.GetProperty("CompletionTokens").GetInt32();
+        }
+        catch (Exception ex) when (ex is KeyNotFoundException)
+        {
+            logger.LogInformation("Usage details not found in model result.");
+        }
+        catch (Exception ex)
+        {
+            logger.LogError(ex, "Error while parsing usage details from model result.");
+            throw;
```

Comment:
Shouldn't this return rather than throw? It's bad for diagnostics to actual introduce failure modes.